### DISCUSSION TITLE
使用typst

### DISCUSSION THREAD

**Discussion by ghost** [TheNetAdmin/zjuthesis#356] [New Ideas]
为什么这个讨论区是空的，好可怕

[typst](https://typst.app/docs/tutorial/)是一群饱受 $LaTeX$ 折磨的人用rust编写的工具。
我想毕业论文的模板也可以用其获得极大的简化，让大家获得更好的体验
所以发邮件询问了作者，以下摘自第二封邮件：

> 找到了[南京大学类似的项目](https://github.com/nju-lug/nju-thesis-typst)[^all-in-one]。
> 他们的格式和我们还是有些区别的。
> 
> 学 typst 可真比学 $\LaTeX$ 简单多了，不过很奇怪的事情是我上次查怎么用 typst 输入 $\prod$ 死活查不到……[^prod]
> 可以用在线工具，也有[各种系统的CLI工具](https://github.com/typst/typst/releases)
> 
> [幻灯片](https://github.com/typst/typst/issues/327#issuecomment-1484076283)在一定程度上也是可以做到的。
> 注意到有这样的操作：
> ```typst
> #text[use text]
> 
> #let te = text
> #te[use te]
> ```
> 故而根据情况选择是否导出为幻灯片应是可行的。
> 
> typst的优点我就不说了，目前发现如下不足：
> 
> - 无法插入PDF，具体可以参看[issue里的讨论](https://github.com/typst/typst/issues/145)，现在要做的话也许需要一下 pdf 转 svg 的脚本。
> - 中文字体的 _italic_ 和 **strong** 还有[问题](https://github.com/typst/typst/issues/635)，我对字体的事情不是很会，今天默认试了一下楷体无法加粗。
> - 无法引用enum的item，感觉做数学的对这个还是有需求的。
> - 字数统计功能我不知
> - 实在要是typst不行的话可以用`pandoc`转成tex
> 
> [^prod]: 后来直接输入了 `$product$` 就好了
> [^all-in-one]: 他们的架构里所有用户的输入内容都在一个 `thesis.typ` 里面，好可怕，对版本管理特别是git很不友好

**Comment by TheNetAdmin** [comment]
我觉得可以先制作一个初版的typest的模版，只针对某个特定模版（比如 undergrad general proposal）看一下效果如何，是否符合学校要求，以及总结一下碰到的问题（例如哪些问题是typest无法解决的）。

之后如果这些都没什么问题，就可以继续搬运latex模版到typest

**Comment by OrangeX4** [comment]
我是 nju-thesis-typst 的作者，很高兴你们也有制作 typst 模板的意愿，如果碰到有任何问题，也可以随时询问我。也推荐加入 793548390 Typst 非官方中文交流群，有许多 typst 的国人 contributors 都在里面，比如 typst-preview 插件的两个作者他们，欢迎一起来交流。

以下是针对你们的疑问的解答：

1. 确实没办法插入 pdf，所以目前只能 pdf 转 svg，可以使用一些在线工具。
2. 伪斜体已经有了相应的 [解决方案](https://typst-doc-cn.github.io/docs/chinese/)，而关于加粗（伪粗体）的问题，群友已经提了相关的 [PR](https://github.com/typst/typst/pull/2970)，并且被 merged 了，相信 typst 的下一个 0.11.0 版本就可以提供相应的伪粗体支持了。
3. 关于引用 enum 的 item 的问题，我不太清楚这个需求的细节，但是我觉得可以通过 show rule 对 enum item 进行改造来解决。
4. 字数统计功能有 [一些 hack 的实现](https://sitandr.github.io/typst-examples-book/book/typstonomicon/word_count.html)，但我不太确定我们是否真的需要这种功能，我是认为一些像 [Word Count CJK](https://marketplace.visualstudio.com/items?itemName=holmescn.vscode-wordcount-cjk) 的插件就能部分解决这个问题。
5. 用 pandoc 转成 tex 也许可行，但是目前我看来 pandoc 的转换效果还是欠佳。

至于全都放在一个 `thesis.typ` 里的设计，这只是我的个人偏好，我个人认为这比较符合 typst 单文件理念，但是并不是强制性的，无论是模板作者还是用户，都可以使用 `#include` 或 `#import` 轻松地切换成多文件架构，可以随意选择你们喜欢的架构。

PS：以及我不太清楚为什么这样会对 git 很不友好，至少我自己用起来感觉还可以（
也许可以说明一下你觉得对 git 这样的版本控制不友好的原因？

**Comment by ghost** [comment]
@OrangeX4 感谢回复！可惜本人不用QQ所以没法加群了

- 我计划用 `pdf2svg` 命令行工具脚本转一下
- 引用 enum item 会出现在某些类似于“(c) 在满足(a)的条件的情况下，证明……”的描述中，后续再试试吧
- 至少我们浙大数学专业还是需要字数统计的[^count]
- 转tex就像双系统的win一样，求个心安，最好不用，确实pandoc还不尽如人意

typst竟然是单文件理念么。不过确乎多文件传递参数什么的挺麻烦

说对git不友好是因为这样的话每次修改论文中任何一个小部分，git都会重新生成thesis.typ的快照。一般来说这个文件会挺大的，那就在存储上就跟每次把word文档拷贝一份保存一样了；如果是SVN这样的管理增量的系统就无所谓
不知道这样解释是否清楚

[^count]: 公式是算一个还是十个字来着……反正图省事就用word自带的字数统计的

**Comment by ghost** [comment]
@TheNetAdmin 总算[开了个头](https://github.com/AmourAmer/zjuthesis/commit/c1369d9ddd8fddad910d354fa4dd4e7af2b5dd35)，以下是目前碰到的一些问题：
- 空白区域的下划线不显示
- 不会读取page的width
- 不会检验文件是否存在

**Reply by OrangeX4** [reply]
1. 空白区域下划线不显示的问题我目前也很头疼，所以只能用 rect 来模拟。
2. 目前无法读取 page 的 width，后续 typst 可能会有 style rework，也许就可以了。
3. 这个目前是预期行为，但是好像有办法 [hack](https://sitandr.github.io/typst-examples-book/book/typstonomicon/try_catch.html)。

**Comment by ghost** [comment]
碰到了一些问题
主要是latex里面用了一些库内置的函数，就不是很知道会导致怎么样的效果;以及typst里全局参数有点点处理起来不优雅
也许需要一个完整的格式规范然后按照架构翻译成文件？
这样也许对某些专业特定的要求的定制也比较友好？[^bb]
不知道以[13年经济学院这个](http://www.cec.zju.edu.cn/2013/0305/c44167a1906138/page.htm)作为出发点好不好
然后我以类似
```
封面 -- cover.typ
  校徽 -- path/to/image
  …
正文 -- body.typ
  字号xx 
…
```
这样的形式写一下（也许用toml格式？），也好明确进度和要求

这样可以么？可以的话我就去创建这样一个格式要求文件，继续做了

[^bb]: 有一说一，个人不是很喜欢重复出现学院的文件夹，也许在zjuthesis/majors/xueyuan里放进需要覆盖的文件比较清晰

**Reply by TheNetAdmin** [reply]
回复得晚了一些， 但我做总结一个现成的规范格式标准

https://github.com/TheNetAdmin/zjuthesis-std

**Reply by ghost** [reply]
抱歉，这些过于复杂，工程量太大了 T^T，~~耻辱销号~~
